File: codeforces/1500~1599/1566/e.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0;
  cin >> n;
  vector<vector<int>> adj(n+1);
  vector<int> degree(n+1);
  for (int i = 1; i < n; ++i) {
    int u = 0, v = 0;
    cin >> u >> v;
    adj[u].push_back(v);
    adj[v].push_back(u);
    degree[u]++;
    degree[v]++;
  }

  vector<bool> is_bud(n+1);
  function<bool(int,int)> dfs = [&](int node, int p) -> bool {
    for (int next_node : adj[node]) {
      if (next_node != p) {
        if (dfs(next_node, node)) {
          degree[node]--;
        }
      }
    }

    bool all_leaf = true;
    for (int next_node : adj[node]) {
      if (next_node != p) {
        if (degree[next_node] != 1 && !is_bud[next_node]) {
          all_leaf = false;
        }
      }
    }

    if (node != 1 && degree[node] > 1 && all_leaf) {
      // cout << "node=" << node << " is bud" << endl;
      return is_bud[node] = true;
    }
    return false;
  };

  dfs(1, -1);

  int bud_count = 0;
  for (int node = 1; node <= n; ++node) {
    if (is_bud[node]) {
      bud_count++;
    }
  }
  // cout << "k=" << bud_count << endl;

  int ans = n - bud_count * 2;
  bool root_has_children = false;
  for (int next_node : adj[1]) {
    if (degree[next_node] == 1) {
      root_has_children = true;
    }
  }

  if (root_has_children) {
    ans--;
  }
  cout << ans << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1500~1599/1566/e.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    if s.is_empty() {
        return Vec::new();
    }
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn dfs(node: usize, p: usize, adj: &Vec<Vec<usize>>, degree: &mut Vec<i32>, is_bud: &mut Vec<bool>) -> bool {
    let mut all_leaf = true;
    for &next_node in adj[node].iter() {
        if next_node == p {
            continue;
        }

        if dfs(next_node, node, adj, degree, is_bud) {
            degree[node] -= 1;
        }
    }

    for &next_node in adj[node].iter() {
        if next_node == p {
            continue;
        }

        if !is_bud[next_node] && degree[next_node] != 1 {
            all_leaf = false;
        }
    }

    if all_leaf && degree[node] > 1 && node != 1 {
        is_bud[node] = true;
        return true;
    }
    false
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());

    let n = read_int::<usize>();
    let mut adj = vec![Vec::new(); n+1];
    let mut degree = vec![0; n+1];
    for _i in 1..n {
        let edge = read_vec::<usize>();
        let u = edge[0];
        let v = edge[1];
        adj[u].push(v);
        adj[v].push(u);
        degree[u] += 1;
        degree[v] += 1;
    }

    let mut is_bud = vec![false; n+1];
    dfs(1, 0, &adj, &mut degree, &mut is_bud);

    let mut bud_count = 0;
    for node in 1..=n {
        if is_bud[node] {
            bud_count += 1;
        }
    }

    let mut root_has_leaf = false;
    for &next_node in adj[1].iter() {
        if degree[next_node] == 1 {
            root_has_leaf = true;
        }
    }

    let mut ans = n - bud_count * 2;
    if root_has_leaf {
        ans -= 1;
    }

    writeln!(writer, "{}", ans).ok();
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
